### PR TITLE
Process workflow notifications

### PR DESCRIPTION
**Pull request by steventux**
https://trello.com/c/0wiO10tC/327-email-trigger-mechanism-for-new-content-tagger-process

Adds a processor to handle non-major-update workflow event notifications.
The current use case for this is sending notifications when content has been re-categorised (re-tagged), the significance being that the content will appear in a high profile collection of finder results.

[See RFC here](https://github.com/alphagov/govuk-rfcs/blob/2cc32ae534d0f30f9702401a31ccd69f73bc4706/rfc-102-add-notification-workflow.md) and [Publishing API PR here](https://github.com/alphagov/publishing-api/pull/1482)

### CODE REVIEW

File: email_alert_service/models/message_processor.rb [hunk 1]
```diff
@@ -1,3 +1,5 @@
+require_relative('./message_processor_validator')
```
Comment (leenagupte) [inline email_alert_service/models/message_processor.rb]: Nitpick: Could this be moved to `major_change_message_processor.rb` as that's where the include is?

File: spec/models/email_alert_spec.rb [hunk 2]
```diff
@@ -89,6 +89,26 @@ def with_lock_unless_done
...
+        email_alert.trigger
+      end
+
+      it "logs receiving a workflow notification for a document" do
+        expect(logger).to have_received(:info).with(
+          "Received workflow message notification for #{document['title']}, with details #{document['details']}"
+        )
+      end
+
+      it "prefers the workflow message to the change note" do
```
Comment (leenagupte) [inline spec/models/email_alert_spec.rb]: Are we expecting to either receive a `change_note` or a `workflow_message`? Is it possible to receive both, and should we test for that?

File: email_alert_service/models/message_processor.rb (hunk 1 above)
Comment (steventux) [reply email_alert_service/models/message_processor.rb]: Good point, needs to be repeated in both processor classes but that's much clearer :+1:

File: spec/models/email_alert_spec.rb (hunk 2 above)
Comment (steventux) [reply spec/models/email_alert_spec.rb]: The test subject does contain both, though it's not that obvious.
The change note is usually built from the change history which is present here https://github.com/alphagov/email-alert-service/blob/719bd65c579350e9fc536d5d19059060137d8f79/spec/models/email_alert_spec.rb#L20 but we prefer the workflow message if present.

**Comment by steventux** [comment]
Thanks @leenagupte that's a fair point about the monstrous first commit, I've amended and attempted to make more atomic changes with slightly more descriptive commit messages.

**Comment by leenagupte** [comment]
> Thanks @leenagupte that's a fair point about the monstrous first commit, I've amended and attempted to make more atomic changes with slightly more descriptive commit messages.

Thanks for splitting things up @steventux. The changes are much easier to follow. 
I think there was just one remaining question about some unused methods in the `WorkflowMessageProcessor` spec, but that other than that, it looks good to me!

File: spec/models/workflow_message_processor_spec.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+  def email_was_not_triggered
+    expect(mock_email_alert).to_not have_received(:trigger)
+  end
+
+  def message_acknowledged
+    expect(channel).to have_received(:acknowledge).with(delivery_tag, false)
+  end
+
+  def message_rejected
```
[the remark this answers is not in the document]
Comment (steventux) [reply spec/models/workflow_message_processor_spec.rb]: Oops missed this - should be sorted now.

**Comment by steventux** [comment]
We don't need to process messages in this way now as there will be direct communication between content-tagger and email-alert-api

**Pull request closed without merging**